core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -19,55 +19,43 @@
 
 package org.apache.iceberg;
 
-import java.util.Collection;
 import java.util.List;
 import java.util.Set;
-import org.apache.iceberg.expressions.Expression;
 import org.apache.iceberg.io.CloseableIterable;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.relocated.com.google.common.collect.FluentIterable;
-import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
 import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
 import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.apache.iceberg.relocated.com.google.common.collect.Sets;
 import org.apache.iceberg.util.SnapshotUtil;
 import org.apache.iceberg.util.ThreadPools;
 
 class IncrementalDataTableScan extends DataTableScan {
-  private long fromSnapshotId;
-  private long toSnapshotId;
-
-  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema, Expression rowFilter,
-                           boolean ignoreResiduals, boolean caseSensitive, boolean colStats,
-                           Collection<String> selectedColumns, ImmutableMap<String, String> options,
-                           long fromSnapshotId, long toSnapshotId) {
-    super(ops, table, null, schema, rowFilter, ignoreResiduals, caseSensitive, colStats, selectedColumns, options);
-    validateSnapshotIds(table, fromSnapshotId, toSnapshotId);
-    this.fromSnapshotId = fromSnapshotId;
-    this.toSnapshotId = toSnapshotId;
+
+  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context.useSnapshotId(null));
+    validateSnapshotIds(table, context.fromSnapshotId(), context.toSnapshotId());
```

Comment:
Shouldn't validations be done in the context? I don't see why the context would allow a change that is invalid and needs to be enforced in different scans.

Reply:
I'm not sure about pushing logic to the `TableScanContext` object, specially validation logic that requires another state for context. I'd agree with validations only for the values of the `snapshotId`s or from within `TableScanContext`, e.g. being different, or non-null, snapshotId not being set if to/from snaphosts are set. However, any other logic dependent of another context like `table` could be specific to where this is being used, for which I'd rather keep `TableScanContext` as simple as possible.

Reply:
I agree with you. If the validation is checking that to and from snapshot IDs are valid, then that makes sense. But for validations like snapshot ID should be null if from snapshot ID is set can be done in the context.